odoo/models.py

```diff
@@ -4946,6 +4946,36 @@ def with_user(self, user):
         """
         return self.with_env(self.env(user=user, su=False))
 
+    def with_company(self, company):
+        """ with_company(company)
+
+        Return a new version of this recordset with a modified context, such that::
+            result.env.company = company
+            result.env.companies = self.env.companies | company
+
+        :raise AccessError: if given company isn't allowed for current user.
+        """
+        if not company:
+            # With company = None/False/0: keep current environment
+            return self
+
+        company_id = int(company)
+        allowed_company_ids = self.env.context.get('allowed_company_ids', [])
+        if allowed_company_ids and company_id == allowed_company_ids[0]:
+            return self
+        # Copy the allowed_company_ids list
+        # to avoid modifying the context of the current environment.
+        allowed_company_ids = list(allowed_company_ids)
+        if company_id in allowed_company_ids:
+            allowed_company_ids.remove(company_id)
+            allowed_company_ids.insert(0, company_id)
+        elif self.env.is_superuser() or company_id in self.env.user.company_ids.ids:
+            allowed_company_ids.insert(0, company_id)
+        else:
+            raise AccessError(_("Company %i isn't allowed for user %s") % (company_id, self.env.user))
```

Comment:
permission checks are redundant with what the env does, let's see `with_company` as syntactic sugar for context manipulation only (and the context is never trusted anyway, even afterwards)